### ISSUE TITLE
Linking Practice Questions

### ISSUE THREAD

**Issue by LEMDO-PM** [projectLEMDO/lemdoIssues#104]
We've had some issues come up with linking practice and would like to confirm and better document best practice for creating links. Objectives of this thread/our linking meeting in January:

1. Determine what encoders expect will happen when their links are rendered vs. what is actually happening when we encode links vs. what is possible to do with the linking mechanisms that we use.
2. Firm up our encoding practice for all types of linking that encoders come across in our project.
3. Create separate GitHub issues for any processing requests that our meeting creates.
4. Create documentation tasks to update our linking chapter based on decisions from this meeting.

Linking questions:

1. When we are linking to a span of text in a modern file from a different file in the same edition (for which we currently use the `<ptr>` element linking to anchors), do we use one or two anchors? What determines how many anchors we point to?
   a. What are the implications on HTML vs. PDF output?
   b. In print, we only really want to use one anchor if we are pointing to something that is one line or less (otherwise we have a span that is line1-line1), but because we don’t generate line numbers until the very end of the process people don’t know whether or not a section will span more than one line.
   c. In our HTML we don't have line numbers, so `<ptr>`s render as the ASSp number. If we have two anchors within the same speech, it renders as Sp.1-Sp.1 (which we don't want), but the section may span more than one line and render as line 1-3 in the PDF output (which we **do** want). How do we reconcile the generated outputs?
2. What is the purpose of `@type="localCit"`? Why don't we use other `@type` values? When exactly do we use it (editions only, within documentation)?
3. What do we expect to happen when people click on a link to a modern text? Does it need to bring them to the exact line, or should it bring them to the start of the relevant speech (i.e., when do we want people to point to a speech xml:id rather than anchors)? Would that fit standard citation conventions in our PDF output?
4. If people link to other editions, does their linking practice change depending on whether or not they are linking to a different anthology?
5. How do we deal with our very important practice of not hard-coding links to the Jenkins site and our need to share individual status pages for the anthologies? JJ to write emails? Who do we tell (and how do we tell them) about where to find these pages?
6. Thinking about new encoders or editors who may not be as tech savvy, how do we clearly explain whether they should be using `<ref>` or `<ptr>`? Here are the ways that people might be linking:
   a. Critical paratext, annotation, or note in a collation to a point in a modern file (same edition).
   b. Critical paratext, annotation, or note in a collation to an act, scene, or speech in a modern file (same edition).
   c. Critical paratext, annotation, or note in a collation to a span in a modern file (same edition).
   d. Character list to a point in a modern file (same file).
   e. Critical paratext, annotation, note in a collation, or character list to a `<div>` in a critical paratext (same edition).
   f. Critical paratext, note in a collation, or character list to an annotation (currently linking to modern file and adding “n.” in the parenthetical citation) (same edition).
   g. Annotation to a different annotation (currently linking to modern file and adding “n.” in the parenthetical citation) (same file).
   h. Critical paratext, annotation, or note in a collation to a point in a modern file (different edition).
   i. Critical paratext, annotation, or note in a collation to an act, scene, or speech in a modern file (different edition).
   j. Critical paratext, annotation, or note in a collation to a span in a modern file (different edition).
   k. Critical paratext, annotation, note in a collation, or character list to a `<div>` in a critical paratext (different edition).
   l. Any file to an external source.
   m. Any file to sitewide databases.
   n. Any file to EEBO, LEME, BEEED, etc.
   o. Documentation to a `<div>` in the same file.
   p. Documentation to a different documentation file.
   q. Documentation to a `<div>` in a different documentation file.
   r. Documentation to Static Search (link currently shows up in diagnostics, but is working on the site).
   s. Documentation to Documentation Index.
   t. Documentation to anthology alpha sites.
   u. Documentation to anthology status pages.
   v. Facsimile links (OS texts, critical paratexts, annotations to `<surface>` elements in facsimile files).
   w. Collations to anchors in the modern text (using `<from>` and `<to>`).
   x. Annotations to anchors in the modern text (using `<target>` and `<targetEnd>`).

Note: This is a lot for encoders to learn and remember. How do we group these/simplify our practice to make it more reasonable for new encoders to learn? Question for Janelle: Is this somewhere that we could start creating what would essentially be lesson plans (as discussed in the grantscrafting meeting)?

**Comment by martindholmes** [comment]
The basic rule for ref versus ptr is:

- If you're linking within an edition, and you want the reference information (act, scene, speech, line(s) etc.) calculated for you, then use ptr.
- If you want to use your own custom text as a link for some reason, or if you are linking to a text which may not be published at the same time or in the same context as the text you're writing in, then use ref. 

The idea of people creating ptr links between completely disparate texts in different editions or anthologies and having the links automagically generate useful linking text is completely impractical, unfortunately. You can only really have autogenerated ASSP/line-based linking within texts which are built and published at the same time as part of the same anthology or PDF. You can't link to another anthology in any way other than a fully hard-coded link to a published edition. Linking to another anthology is the same as linking to any external website. So if an anthology has never been published, or a text you want to link to hasn't yet been published in its anthology, then you can't link to it; end of story. And when you do link, you will have to link to a specific edition using the full edition URL, not to the current edition whatever it is, because of course the text you're linking to may change in a future edition.

On the issue of Jenkins links, I don't see a problem. Everyone working on texts will need to know where the Jenkins builds are, so that they can see their texts; they'll also need to be warned not to share such URLs.  The only people who need to see anthology status pages are anthology editors, and they already know where their anthology is. Documentation may not contain hard-coded links to build products on the Jenkins server, period; since documentation will be published on the LEMDO site, it may only ever link to public resources. Anthology editors are fully capable of bookmarking a page in their browser. 

Many of the scenarios in #6 are already covered by our standard linking documentation, surely? To link to a person, for instance, you're going to use `<name>` or `<persName>` (see issue #2) with `@ref` and an appropriate prefix; you're not going to use `<ref>` or `<ptr>`; ditto for any of the sitewide collections. Some of the other scenarios, such as linking to facsimile pages, currently have no protocol and have never had one; before we can create one, we will need to decide whether facsimiles/page-image surrogates are part of editions or not. If they are, then we have to duplicate the complete set of page-images for every edition we publish of a text; on the other hand, we could treat our collection of facsimiles as an entirely separate project, into which any edition can link, and which we host separately at (for instance) lemdo.uvic.ca/facsimiles. In that case, linking would be the same as linking to any other external resource.

**Comment by LEMDO-PM** [comment]
- Linking to lemma in modern text from annotations-only pages
- Linking to lemma in modern text from collations-only pages

**Comment by JanelleJenstad** [comment]
@Navarra-Houldin : Agreed that we need lesson plans on linking. Let's plan for a workshop for LEMDO editors.

**Comment by LEMDO-PM** [comment]
Notes:
- What does "local" in localCit mean? Must be careful because we only want to link to things that are published alongside yours. It is intended to only be used within an edition. Add documentation, write Schematron that `<ptr>` always have @localCit when in an edition file (i.e., not documentation) and not in `<linkGrp>`.
- Change Number Figures to Give Figures an xml:id in documentation. xml:ids will allow us to link to figures. We can't add figure numbers in the `<figDesc>` as it would mess up the PDF auto-numbering.
- Only link to things that are already published.
- In documentation, we use `<ptr>`, but don't need `@type="localCit"`
- When linking to edition in a different anthology, must do `<ref target="published URL including _editions/#.#">`. E.g., moms_editions/1.0 Discuss citation for this (must always include version number).
- For anthology status pages, our menus are lemdo-dev specific (and different from lemdo menus). Allow hard-coded link in lemdo-dev menus, do not allow them in any documentation or files. @JanelleJenstad, please add links back
- If it is within your own edition, use `<ptr>` for auto-generated and `<ref>` for choosing your own rendered text. Use `<ref>` outside of your own edition.
- @martindholmes please revisit [Issue 92](https://github.com/projectLEMDO/lemdoIssues/issues/92)
- Write Schematron to require `<head>` as a child of `<div>` when the `<div>` has an xml:id.
- Create ticket for Martin to add processing to link to lemma in modern text from annotations and collations page.

**Comment by martindholmes** [comment]
Can this be closed now? It seems that each of the above actions, if still relevant, deserves its own ticket. Issues with more than one action will get neglected.

**Comment by LEMDO-PM** [comment]
Yes. Everything that needs its own ticket has one. Closing now.
